criu/cr-restore.c

```diff
@@ -1107,21 +1107,22 @@ static int restore_one_zombie(CoreEntry *core)
 
 static int setup_newborn_fds(struct pstree_item *me)
 {
-	if (clone_service_fd(me))
+	bool cleanup_parent_sfdt = !me->parent ||
+				   (rsti(me->parent)->fdt &&
+				    !(rsti(me)->clone_flags & CLONE_FILES));
+
+	if (clone_service_fd(me, cleanup_parent_sfdt))
```

Comment:
I understand what you want to do in this patch, but this looks like a hack.

I think we can make it simpler and write a comment that other descriptors will be cleaned up in close_old_fds.

```
diff --git a/criu/servicefd.c b/criu/servicefd.c
index 1661e12af..89138028f 100644
--- a/criu/servicefd.c
+++ b/criu/servicefd.c
@@ -294,6 +294,14 @@ int clone_service_fd(struct pstree_item *me)
        if (service_fd_base == new_base && service_fd_id == id)
                return 0;
 
+
+       {
+               int fd = __get_service_fd(PROC_SELF_FD_OFF, id);
+               close(fd);
+               clear_bit(PROC_SELF_FD_OFF, sfd_map);
+       }
+
+
        /* Dup sfds in memmove() style: they may overlap */
        if (get_service_fd(LOG_FD_OFF) < new_base - LOG_FD_OFF - SERVICE_FD_MAX * id)
                for (i = SERVICE_FD_MIN + 1; i < SERVICE_FD_MAX; i++)

```

Reply:
we call clone_service_fd in a new process and we have to close PROC_SELF_FD_OFF anyway.